realm/realm-library/src/main/java/io/realm/internal/ColumnInfo.java

```diff
@@ -25,13 +25,14 @@
 
 
 /**
- * Objects of this class play two roles:
- * <ul>
- * <li>Subclasses are a fast cache of column indices, for proxy object</li>
- * <li>They cache table (schema) information used by for StandardRealmObjectSchema and StandardRealmSchema</li>
- * </ul>
+ * The Subclasses of this are a fast cache of column indices, for proxy object.
+ * <p>
  * The fast cache functionality is implemented in the Proxy classes generated by {@code RealmProxyClassGenerator}.
- * Be sure to understand what is going on there, before changing things here.
+ * Every proxy object will hold an reference of its {@code ColumnInfo}. The ref to the {@code ColumnInfo} instance is
+ * maintained by the {@link ColumnIndices}. As long as the reference in proxy object has been set, it should never be
+ * changed. When schema changes, the relevant {@code ColumnInfo} instance's content will be refreshed therefore the
+ * proxy object could have the latest column indices. Be sure to understand what is going on there,
+ * before changing things here.
```

Comment:
Because these things are passed between threads.  Some of them have to be mutable.  Some of them absolutely cannot be.